envoy/cilium_network_policy.h

```diff
@@ -0,0 +1,218 @@
+#pragma once
+
+#include "envoy/local_info/local_info.h"
+#include "envoy/upstream/cluster_manager.h"
+#include "envoy/event/dispatcher.h"
+
+#include "common/common/logger.h"
+#include "common/router/config_utility.h"
+#include "envoy/config/subscription.h"
+#include "envoy/singleton/instance.h"
+#include "envoy/thread_local/thread_local.h"
+#include "envoy/http/header_map.h"
+
+#include "cilium/npds.pb.h"
+
+namespace Envoy {
+namespace Cilium {
+
+class NetworkPolicyMap : public Singleton::Instance,
+                         Config::SubscriptionCallbacks<cilium::NetworkPolicy>,
+                         public Logger::Loggable<Logger::Id::config> {
+public:
+  NetworkPolicyMap(const envoy::api::v2::core::ApiConfigSource& api_config_source,
+		   const LocalInfo::LocalInfo& local_info,
+		   Upstream::ClusterManager& cm, Event::Dispatcher& dispatcher,
+		   Stats::Scope &scope, ThreadLocal::SlotAllocator& tls);
+  NetworkPolicyMap(std::unique_ptr<Envoy::Config::Subscription<cilium::NetworkPolicy>>&& subscription,
+		   ThreadLocal::SlotAllocator& tls);
+  ~NetworkPolicyMap() {}
+
+  class PolicyInstance {
+  public:
+    PolicyInstance(uint64_t hash, const cilium::NetworkPolicy& proto)
+        : hash_(hash), policy_proto_(proto), ingress_(policy_proto_.ingress_per_port_policies()),
+          egress_(policy_proto_.egress_per_port_policies()) {}
+
+    uint64_t hash_;
+    const cilium::NetworkPolicy policy_proto_;
+
+  protected:
+    class HttpNetworkPolicyRule : public Logger::Loggable<Logger::Id::config> {
+    public:
+      HttpNetworkPolicyRule(const cilium::HttpNetworkPolicyRule& rule) {
+	ENVOY_LOG(trace, "Cilium L7 HttpNetworkPolicyRule():");
+	for (const auto& header: rule.headers()) {
+	  headers_.emplace_back(header);
+	  const auto& header_data = headers_.back();
+	  ENVOY_LOG(trace, "Cilium L7 HttpNetworkPolicyRule(): HeaderData {}={} ({})",
+		    header_data.name_.get(), header_data.value_, header_data.is_regex_ ? "regex" : "literal");
+	}
+      }
+
+      bool Matches(const Envoy::Http::HeaderMap& headers) const {
+	// Empty set matches any headers.
+	return Envoy::Router::ConfigUtility::matchHeaders(headers, headers_);
+      }
+
+      std::vector<Envoy::Router::ConfigUtility::HeaderData> headers_; // Allowed if empty.
+    };
+    
+    class PortNetworkPolicyRule : public Logger::Loggable<Logger::Id::config> {
+    public:
+      PortNetworkPolicyRule(const cilium::PortNetworkPolicyRule& rules) {
+	for (const auto& remote: rules.remote_policies()) {
+	  ENVOY_LOG(trace, "Cilium L7 PortNetworkPolicyRule(): Allowing remote {}", remote);
+	  allowed_remotes_.emplace(remote);
+	}
+	for (const auto& http_rule: rules.http_rules().http_rules()) {
```

Comment:
You must check that `rules.http_rules()` is set before doing that loop.
This can be the case when mixing L4-only rules and L4+L7 rules in the same policy.

Reply:
Thanks! Added check for `rules.has_http_rules()` before the loop.